nifty-ssl/src/main/java/com/facebook/nifty/ssl/ThreadLocalSslBufferPool.java

```diff
@@ -0,0 +1,139 @@
+ */
+package com.facebook.nifty.ssl;
+
+import org.jboss.netty.handler.ssl.SslBufferPool;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+import static com.facebook.nifty.core.NettyConfigBuilderBase.DEFAULT_WORKER_THREAD_COUNT;
+
+public class ThreadLocalSslBufferPool extends SslBufferPool {
+    private final ThreadLocal<NonBlockingSslBufferPoll> pool;
+
+    public ThreadLocalSslBufferPool(int maxPoolSize, boolean preallocate, boolean allocateDirect) {
+        pool = new ThreadLocal<NonBlockingSslBufferPoll>() {
+            @Override
+            protected NonBlockingSslBufferPoll initialValue() {
+                int threadLocalMaxPoolSize = maxPoolSize / DEFAULT_WORKER_THREAD_COUNT;
+                if (threadLocalMaxPoolSize % DEFAULT_WORKER_THREAD_COUNT != 0) {
+                    threadLocalMaxPoolSize ++;
+                }
+                return new NonBlockingSslBufferPoll(threadLocalMaxPoolSize, preallocate, allocateDirect);
+            }
+        };
+    }
+
+    @Override
+    public int getMaxPoolSize() {
+        return pool.get().getMaxPoolSize();
+    }
+
+    @Override
+    public int getUnacquiredPoolSize() {
+        return pool.get().getUnacquiredPoolSize();
+    }
+
+    @Override
+    public ByteBuffer acquireBuffer() {
+        return pool.get().acquireBuffer();
+    }
+
+    @Override
+    public void releaseBuffer(ByteBuffer buffer) {
+        pool.get().releaseBuffer(buffer);
+    }
+
+    /**
+     * A lot of this code is taken from Netty's OpenSslServerContext.
+     * https://github.com/netty/netty/blob/3.10/src/main/java/org/jboss/netty/handler/ssl/SslBufferPool.java
+     * If we use threadLocal SslBufferPool with a maxPoolSize limit and the thread runs out of preallocated buffer
+     * space, the thread will deadlock. UnlimitedSslBufferPool allows us to create a SslBufferPool without a maxPoolSize
+     * limit to avoid this deadlock.
+     */
+    public static class NonBlockingSslBufferPoll {
+        private static final int MAX_PLAINTEXT_LENGTH = 16 * 1024; // 2^14
+        private static final int MAX_COMPRESSED_LENGTH = MAX_PLAINTEXT_LENGTH + 1024;
+        private static final int MAX_CIPHERTEXT_LENGTH = MAX_COMPRESSED_LENGTH + 1024;
+
+        // Header (5) + Data (2^14) + Compression (1024) + Encryption (1024) + MAC (20) + Padding (256)
+        private static final int MAX_ENCRYPTED_PACKET_LENGTH = MAX_CIPHERTEXT_LENGTH + 5 + 20 + 256;
+
+        // Add 1024 as a room for compressed data and another 1024 for Apache Harmony compatibility.
+        private static final int MAX_PACKET_SIZE_ALIGNED = (MAX_ENCRYPTED_PACKET_LENGTH / 128 + 1) * 128;
+
+        private final BlockingQueue<ByteBuffer> pool;
+        private final int maxBufferCount;
+        private final boolean allocateDirect;
+
+        public NonBlockingSslBufferPoll(int maxPoolSize, boolean preallocate, boolean allocateDirect) {
```

Comment:
this needs some unit tests for different cases, some ideas are:

1. when maxPoolSize way lower than requested buffers, that this does not block
2. Size of the buffer returned is correct